### PR TITLE
Use Paginator to display results

### LINKED ISSUE #1360: Fetch join queries do not work on `3.31`
#### Sonata packages

<details><summary>show</summary>
<p>

```
$ composer show --latest 'sonata-project/*'
sonata-project/admin-bundle        3.93.0 3.93.0 The missing Symfony Admin Generator
sonata-project/block-bundle        4.5.2  4.5.2  Symfony SonataBlockBundle
sonata-project/cache               2.1.1  2.1.1  Cache library
sonata-project/doctrine-extensions 1.12.0 1.12.0 Doctrine2 behavioral extensions
sonata-project/exporter            2.5.1  2.5.1  Lightweight Exporter library
sonata-project/form-extensions     1.9.0  1.9.0  Symfony form extensions
sonata-project/twig-extensions     1.5.1  1.5.1  Sonata twig extensions
```

</p>
</details>

#### Symfony packages

<details><summary>show</summary>
<p>

```
$ composer show --latest 'symfony/*'
symfony/asset                      v5.2.4  v5.2.4  Manages URL generation and versioning of web assets such as CSS stylesheets, JavaScript files and image files
symfony/browser-kit                v5.2.4  v5.2.4  Simulates the behavior of a web browser, allowing you to make requests, click on links and submit forms programmatically
symfony/cache                      v5.2.4  v5.2.4  Provides an extended PSR-6, PSR-16 (and tags) implementation
symfony/cache-contracts            v2.2.0  v2.2.0  Generic abstractions related to caching
symfony/config                     v5.2.4  v5.2.4  Helps you find, load, combine, autofill and validate configuration values of any kind
symfony/console                    v4.4.20 v5.2.5  Eases the creation of beautiful and testable command line interfaces
symfony/debug                      v4.4.20 v4.4.20 Provides tools to ease debugging PHP code
symfony/dependency-injection       v5.2.5  v5.2.5  Allows you to standardize and centralize the way objects are constructed in your application
symfony/deprecation-contracts      v2.2.0  v2.2.0  A generic function and convention to trigger deprecation notices
symfony/doctrine-bridge            v4.4.20 v5.2.5  Provides integration for Doctrine with various Symfony components
symfony/dom-crawler                v5.2.4  v5.2.4  Eases DOM navigation for HTML and XML documents
symfony/error-handler              v4.4.20 v5.2.4  Provides tools to manage errors and ease debugging PHP code
symfony/event-dispatcher           v4.4.20 v5.2.4  Provides tools that allow your application components to communicate with each other by dispatching events and listening to them
symfony/event-dispatcher-contracts v1.1.9  v2.2.0  Generic abstractions related to dispatching event
symfony/expression-language        v5.2.4  v5.2.4  Provides an engine that can compile and evaluate expressions
symfony/filesystem                 v5.2.4  v5.2.4  Provides basic utilities for the filesystem
symfony/finder                     v5.2.4  v5.2.4  Finds files and directories via an intuitive fluent interface
symfony/form                       v4.4.20 v5.2.5  Allows to easily create, process and reuse HTML forms
symfony/framework-bundle           v4.4.20 v5.2.5  Provides a tight integration between Symfony components and the Symfony full-stack framework
symfony/http-client                v5.2.4  v5.2.4  Provides powerful methods to fetch HTTP resources synchronously or asynchronously
symfony/http-client-contracts      v2.3.1  v2.3.1  Generic abstractions related to HTTP clients
symfony/http-foundation            v5.2.4  v5.2.4  Defines an object-oriented layer for the HTTP specification
symfony/http-kernel                v4.4.20 v5.2.5  Provides a structured process for converting a Request into a Response
symfony/intl                       v5.2.4  v5.2.4  Provides a PHP replacement layer for the C intl extension that includes additional data from the ICU library
symfony/options-resolver           v5.2.4  v5.2.4  Provides an improved replacement for the array_replace PHP function
symfony/panther                    v1.0.1  v1.0.1  A browser testing and web scraping library for PHP and Symfony.
symfony/phpunit-bridge             v5.2.4  v5.2.4  Provides utilities for PHPUnit, especially user deprecation notices management
symfony/polyfill-ctype             v1.22.1 v1.22.1 Symfony polyfill for ctype functions
symfony/polyfill-intl-grapheme     v1.22.1 v1.22.1 Symfony polyfill for intl's grapheme_* functions
symfony/polyfill-intl-icu          v1.22.1 v1.22.1 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-intl-normalizer   v1.22.1 v1.22.1 Symfony polyfill for intl's Normalizer class and related functions
symfony/polyfill-mbstring          v1.22.1 v1.22.1 Symfony polyfill for the Mbstring extension
symfony/polyfill-php72             v1.22.1 v1.22.1 Symfony polyfill backporting some PHP 7.2+ features to lower PHP versions
symfony/polyfill-php73             v1.22.1 v1.22.1 Symfony polyfill backporting some PHP 7.3+ features to lower PHP versions
symfony/polyfill-php80             v1.22.1 v1.22.1 Symfony polyfill backporting some PHP 8.0+ features to lower PHP versions
symfony/process                    v5.2.4  v5.2.4  Executes commands in sub-processes
symfony/property-access            v5.2.4  v5.2.4  Provides functions to read and write from/to an object or array using a simple string notation
symfony/property-info              v5.2.4  v5.2.4  Extracts information about PHP class' properties using metadata of popular sources
symfony/routing                    v5.2.4  v5.2.4  Maps an HTTP request to a set of configuration variables
symfony/security-acl               v3.1.1  v3.1.1  Symfony Security Component - ACL (Access Control List)
symfony/security-bundle            v4.4.20 v5.2.5  Provides a tight integration of the Security component into the Symfony full-stack framework
symfony/security-core              v4.4.20 v5.2.5  Symfony Security Component - Core Library
symfony/security-csrf              v5.2.4  v5.2.4  Symfony Security Component - CSRF Library
symfony/security-guard             v4.4.20 v5.2.4  Symfony Security Component - Guard
symfony/security-http              v4.4.20 v5.2.5  Symfony Security Component - HTTP Integration
symfony/service-contracts          v2.2.0  v2.2.0  Generic abstractions related to writing services
symfony/string                     v5.2.4  v5.2.4  Provides an object-oriented API to strings and deals with bytes, UTF-8 code points and grapheme clusters in a unified way
symfony/templating                 v5.2.4  v5.2.4  Provides all the tools needed to build any kind of template system
symfony/translation                v4.4.20 v5.2.5  Provides tools to internationalize your application
symfony/translation-contracts      v2.3.0  v2.3.0  Generic abstractions related to translation
symfony/twig-bridge                v4.4.20 v5.2.5  Provides integration for Twig with various Symfony components
symfony/twig-bundle                v4.4.20 v5.2.4  Provides a tight integration of Twig into the Symfony full-stack framework
symfony/validator                  v5.2.5  v5.2.5  Provides tools to validate values
symfony/var-dumper                 v5.2.5  v5.2.5  Provides mechanisms for walking through any arbitrary PHP variable
symfony/var-exporter               v5.2.4  v5.2.4  Allows exporting any serializable PHP data structure to plain PHP code
symfony/yaml                       v5.2.5  v5.2.5  Loads and dumps YAML files
```

</p>
</details>

#### PHP version

```
$ php -v
PHP 7.4.14 (cli) (built: Jan  8 2021 13:20:04) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.14, Copyright (c), by Zend Technologies
    with Xdebug v2.9.2, Copyright (c) 2002-2020, by Derick Rethans
```

## Subject

Version 3.31 (https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/1319) introduced a BC break when using a fetch join query in `AdminInterface::configureQuery`, when the query is executed, the object is not fully hydrated, it only has one item in a one-to-many association.

## Minimal repository with the bug

I created a test here https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/1354 that it works on `3.30`.

## Steps to reproduce

It's already in the test, having an entity:
```php
class Author
{
    /**
     * @ORM\OneToMany(targetEntity=Book::class, mappedBy="author")
     *
     * @var Collection<array-key, Book>
     */
    private $books;
}
```
in `AuthorAdmin`:

```php
protected function configureQuery(ProxyQueryInterface $query): ProxyQueryInterface
{
    $alias = $query->getQueryBuilder()->getRootAliases()[0];

    $query
        ->addSelect('book')
        ->leftJoin(sprintf('%s.books', $alias), 'book');

    return $query;
}
```

Then adding 2 books to an author, then go to list page.

## Expected results

The number of books should be 2.

image

## Actual results

image

### LINKED ISSUE #1361: Paginator broken when using entity inheritance
#### Sonata packages

<details><summary>show</summary>
<p>

```
$ composer show --latest 'sonata-project/*'
sonata-project/admin-bundle              3.93.0 3.93.0 The missing Symfony Admin Generator
sonata-project/block-bundle              4.5.2  4.5.2  Symfony SonataBlockBundle
sonata-project/cache                     2.1.1  2.1.1  Cache library
sonata-project/doctrine-extensions       1.12.0 1.12.0 Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.31.0 3.31.0 Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/exporter                  2.5.1  2.5.1  Lightweight Exporter library
sonata-project/form-extensions           1.9.0  1.9.0  Symfony form extensions
sonata-project/twig-extensions           1.5.1  1.5.1  Sonata twig extensions
```

</p>
</details>

#### Symfony packages

<details><summary>show</summary>
<p>

```
$ composer show --latest 'symfony/*'
symfony/asset                      v4.4.20 v5.2.4  Manages URL generation and versioning of web assets such as CSS stylesheets, JavaScript files and image files
symfony/cache                      v4.4.20 v5.2.4  Provides an extended PSR-6, PSR-16 (and tags) implementation
symfony/cache-contracts            v2.2.0  v2.2.0  Generic abstractions related to caching
symfony/config                     v4.4.20 v5.2.4  Helps you find, load, combine, autofill and validate configuration values of any kind
symfony/console                    v4.4.20 v5.2.5  Eases the creation of beautiful and testable command line interfaces
symfony/debug                      v4.4.20 v4.4.20 Provides tools to ease debugging PHP code
symfony/dependency-injection       v4.4.20 v5.2.5  Allows you to standardize and centralize the way objects are constructed in your application
symfony/doctrine-bridge            v4.4.20 v5.2.5  Provides integration for Doctrine with various Symfony components
symfony/dotenv                     v4.4.20 v5.2.4  Registers environment variables from a .env file
symfony/error-handler              v4.4.20 v5.2.4  Provides tools to manage errors and ease debugging PHP code
symfony/event-dispatcher           v4.4.20 v5.2.4  Provides tools that allow your application components to communicate with each other by dispatching events and listening to them
symfony/event-dispatcher-contracts v1.1.9  v2.2.0  Generic abstractions related to dispatching event
symfony/expression-language        v4.4.20 v5.2.4  Provides an engine that can compile and evaluate expressions
symfony/filesystem                 v4.4.20 v5.2.4  Provides basic utilities for the filesystem
symfony/finder                     v4.4.20 v5.2.4  Finds files and directories via an intuitive fluent interface
symfony/flex                       v1.12.2 v1.12.2 Composer plugin for Symfony
symfony/form                       v4.4.20 v5.2.5  Allows to easily create, process and reuse HTML forms
symfony/framework-bundle           v4.4.20 v5.2.5  Provides a tight integration between Symfony components and the Symfony full-stack framework
symfony/http-client-contracts      v2.3.1  v2.3.1  Generic abstractions related to HTTP clients
symfony/http-foundation            v4.4.20 v5.2.4  Defines an object-oriented layer for the HTTP specification
symfony/http-kernel                v4.4.20 v5.2.5  Provides a structured process for converting a Request into a Response
symfony/inflector                  v4.4.20 v5.2.4  Converts words between their singular and plural forms (English only)
symfony/intl                       v4.4.20 v5.2.4  Provides a PHP replacement layer for the C intl extension that includes additional data from the ICU library
symfony/mime                       v4.4.20 v5.2.5  Allows manipulating MIME messages
symfony/options-resolver           v4.4.20 v5.2.4  Provides an improved replacement for the array_replace PHP function
symfony/polyfill-intl-grapheme     v1.22.1 v1.22.1 Symfony polyfill for intl's grapheme_* functions
symfony/polyfill-intl-icu          v1.22.1 v1.22.1 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-intl-idn          v1.22.1 v1.22.1 Symfony polyfill for intl's idn_to_ascii and idn_to_utf8 functions
symfony/polyfill-intl-normalizer   v1.22.1 v1.22.1 Symfony polyfill for intl's Normalizer class and related functions
symfony/polyfill-mbstring          v1.22.1 v1.22.1 Symfony polyfill for the Mbstring extension
symfony/polyfill-php72             v1.22.1 v1.22.1 Symfony polyfill backporting some PHP 7.2+ features to lower PHP versions
symfony/polyfill-php73             v1.22.1 v1.22.1 Symfony polyfill backporting some PHP 7.3+ features to lower PHP versions
symfony/polyfill-php80             v1.22.1 v1.22.1 Symfony polyfill backporting some PHP 8.0+ features to lower PHP versions
symfony/property-access            v4.4.20 v5.2.4  Provides functions to read and write from/to an object or array using a simple string notation
symfony/routing                    v4.4.20 v5.2.4  Maps an HTTP request to a set of configuration variables
symfony/security-acl               v3.1.1  v3.1.1  Symfony Security Component - ACL (Access Control List)
symfony/security-bundle            v4.4.20 v5.2.5  Provides a tight integration of the Security component into the Symfony full-stack framework
symfony/security-core              v4.4.20 v5.2.5  Symfony Security Component - Core Library
symfony/security-csrf              v4.4.20 v5.2.4  Symfony Security Component - CSRF Library
symfony/security-guard             v4.4.20 v5.2.4  Symfony Security Component - Guard
symfony/security-http              v4.4.20 v5.2.5  Symfony Security Component - HTTP Integration
symfony/service-contracts          v2.2.0  v2.2.0  Generic abstractions related to writing services
symfony/stopwatch                  v4.4.20 v5.2.4  Provides a way to profile code
symfony/string                     v5.2.4  v5.2.4  Provides an object-oriented API to strings and deals with bytes, UTF-8 code points and grapheme clusters in a unified way
symfony/translation                v4.4.20 v5.2.5  Provides tools to internationalize your application
symfony/translation-contracts      v2.3.0  v2.3.0  Generic abstractions related to translation
symfony/twig-bridge                v4.4.20 v5.2.5  Provides integration for Twig with various Symfony components
symfony/twig-bundle                v4.4.20 v5.2.4  Provides a tight integration of Twig into the Symfony full-stack framework
symfony/validator                  v4.4.20 v5.2.5  Provides tools to validate values
symfony/var-dumper                 v4.4.20 v5.2.5  Provides mechanisms for walking through any arbitrary PHP variable
symfony/var-exporter               v4.4.20 v5.2.4  Allows exporting any serializable PHP data structure to plain PHP code
symfony/web-profiler-bundle        v4.4.20 v5.2.4  Provides a development tool that gives detailed information about the execution of any request
symfony/yaml                       v4.4.20 v5.2.5  Loads and dumps YAML files
```

</p>
</details>

#### PHP version

```
$ php -v
PHP 8.0.3 (cli) (built: Mar  6 2021 03:11:06) ( NTS )
Copyright (c) The PHP Group
Zend Engine v4.0.3, Copyright (c) Zend Technologies
    with Zend OPcache v8.0.3, Copyright (c), by Zend Technologies
    with Xdebug v3.0.3, Copyright (c) 2002-2021, by Derick Rethans

```

## Subject

Since v3.31 listing entities which extend another entities is broken. The generated count query uses GROUP BY (since https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/1319 ) but the non-aggregated fields are not part of the group by statement.

Reverting to v3.30 solves the issue.

## Minimal repository with the bug

https://github.com/Devristo/sonata-admin-entity-inheritance

## Steps to reproduce

```
git clone git@github.com:Devristo/sonata-admin-entity-inheritance.git sonata-admin-entity-inheritance
cd sonata-admin-entity-inheritance
docker-compose up -d
bin/console doctrine:schema:update --force
bin/console doctrine:fixtures:load -n
symfony serve
```
Then go to: https://127.0.0.1:8000/admin/app/sub/list

## Expected results

You will see a listing.

## Actual results
```

An exception has been thrown during the rendering of a template ("An exception occurred while executing 'SELECT b0_.id AS id_0, s1_.other_field AS other_field_1, b0_.discr AS discr_2 FROM sub s1_ INNER JOIN base b0_ ON s1_.id = b0_.id GROUP BY b0_.id ORDER BY b0_.id ASC LIMIT 32':

SQLSTATE[42803]: Grouping error: 7 ERROR: column "s1_.other_field" must appear in the GROUP BY clause or be used in an aggregate function
LINE 1: SELECT b0_.id AS id_0, s1_.other_field AS other_field_1, b0_...
^").
```

### PR DESCRIPTION
**Pull request by VincentLanglet**
## Subject

I am targeting this branch, because bugfix.

Closes #1360.
Closes #1361.

## Changelog

```markdown
### Fixed
- Fetch join queries for Pager
- Pager when using entity inheritance
```

### CODE REVIEW

**Comment by VincentLanglet** [comment]
Current error: `An exception has been thrown during the rendering of a template ("Single id is not allowed on composite primary key in entity Sonata\DoctrineORMAdminBundle\Tests\App\Entity\Car").`

**Comment by VincentLanglet** [comment]
@Warxcell You introduce the Paginator in order to fix an issue with composite keys.
But I currently get the exact opposite. Do you know how to solve my issue ?

**Comment by Warxcell** [comment]
How that passed tests when merged? :thinking:

**Comment by VincentLanglet** [comment]
> How that passed tests when merged and now it's failing? 🤔

Paginator was working fine for the count.
Now I try to use it for the `getCurrentResults` too in order to have consistency between the count and the results, and in order to fix some others issues.

But for the results, I got
```
An exception has been thrown during the rendering of a template ("Single id is not allowed on composite primary key in entity Sonata\DoctrineORMAdminBundle\Tests\App\Entity\Car").
```
And discover https://github.com/doctrine/orm/issues/2910.
I was wondering if you knew some things about this issue.

And since you're using composite primary key, it would be nice to try my code on your project in order to check it doesn't introduce any bug.

**Comment by Warxcell** [comment]
Ok, got it. Will try with your branch.

**Comment by VincentLanglet** [comment]
> Ok, got it. Will try with your branch.
> But `1 === \count($identifierFieldNames)` is wrong as second parameter. This will not work when ToMany Collection is joined.

With fetchJoinCollection flag set to true, it throws an error for composite primary keys
cf: https://github.com/doctrine/orm/issues/2910, especially the last comment.

> In order to use Paginator with a composite primary key entity, you need to instantiate the Paginator with the $fetchJoinCollection flag set to false (it defaults to true). Now, pagination will skip the WhereInWalker which throws an exception when using a composite primary key. Fetch joins with entities (one-to-one or many-to-one) will still work. You can even use a regular join with a collection.
>
> The only "problem" is when your Paginator query calls for a fetch join to a collection. The work around for this is to use a regular join as mentioned above. The drawback is that your paginated entities will not be hydrated with the collection. The collection will have to be lazy-loaded when called upon.

So JOIN should still works. You just cannot use the benefit of fetch-join.

**Comment by Warxcell** [comment]
So it's basically it's not supported to have composite PK and join To Many Collection.

**Comment by Warxcell** [comment]
Just tried out your branch - on my side Paginator with second argument `true` generates wrong query at all. 
Tracking it down now.
With your check `1 === \count($identifierFieldNames)` which results in `false` - it's working fine, but I didn't tried it  if I join ToMany Association, must create one artificially to test that.

**Comment by Warxcell** [comment]
On my case it's because of ORDER. Doctrine doesn't generate correct aliases when using order by composite key (or maybe even single id ?)

```
SELECT DISTINCT product_id_2, channel_id_3
FROM (SELECT DISTINCT product_id_2, channel_id_3, s0_.product_id, s0_.channel_id
      FROM (SELECT s0_.created    AS created_0,
                   s0_.amount     AS amount_1,
                   s0_.product_id AS product_id_2,
                   s0_.channel_id AS channel_id_3
            FROM supplier_product_our_retail_prices s0_) dctrn_result_inner
      ORDER BY s0_.product_id ASC, s0_.channel_id ASC) dctrn_result
LIMIT 32
```


this is the generated query.
1. problem is that product_id is selected twice, once with correct alias, once with wrong ones, and order by is with wrong aliases.

**Comment by Warxcell** [comment]
I guess we can propose a change to Doctrine Paginator to support composite PKs and generate queries like:

`WHERE pk1 IN(:pk1) AND pk2 IN(:pk2)...` ?

**Comment by VincentLanglet** [comment]
> After I fix that, I arrive to `An exception has been thrown during the rendering of a template ("Single id is not allowed on composite primary key in entity ").`

Even with the fetchAssociation = false ?

> I guess we can propose a change to Doctrine Paginator to support composite PKs and generate queries like:
> 
> `WHERE pk1 IN(:pk1) AND pk2 IN(:pk2)...` ?

Sure, if you know how to fix it.

Can you also try the commit bbfda252e2396f68b250bee14fcb58969a045122 @Warxcell ?
This is another workaround who work for the tests. I just don't use Paginator if it's a composite primary key ^^'

**Comment by Warxcell** [comment]
With `fetchAssociation = false` it's ok, no error.

**Comment by greg0ire** [comment]
Can you please add tests for the issues you fixed? It would be great to check that these tests also pass without #1319 , to make sure there is no other breaking change.

**Comment by greg0ire** [comment]
Ah there is one here: https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/1354/commits/f44aa0f5948e9af108d6de2cadef3b66357889b7 Let's cherry-pick it?

**Comment by franmomu** [comment]
> Can you please add tests for the issues you fixed? It would be great to check that these tests also pass without #1319 , to make sure there is no other breaking change.

We definitely should, I created https://github.com/sonata-project/SonataDoctrineORMAdminBundle/issues/1370, but we should add it before this.

Based on https://github.com/sonata-project/SonataDoctrineORMAdminBundle/issues/1360#issuecomment-805763483, should we test on different platforms? I think it would be interesting, it will also prevent things like https://github.com/sonata-project/SonataDoctrineORMAdminBundle/issues/1233

**Comment by VincentLanglet** [comment]
I cherry picked the tests from @franmomu so the PR is RTM @sonata-project/contributors

**Review by jordisala1991 — DISMISSED** [review]

File: tests/App/Admin/AuthorAdmin.php [hunk 1]
```diff
@@ -25,7 +29,17 @@ protected function configureListFields(ListMapper $list): void
...
+            ->add('number_of_books', FieldDescriptionInterface::TYPE_INTEGER, [
+                'accessor' => static function (Author $author): int {
+                    return $author->getBooks()->count();
+                },
+                'template' => 'author/list_number_of_books_field.html.twig',
+            ])
+            ->add('numberOfReaders', FieldDescriptionInterface::TYPE_INTEGER, [
+                'template' => 'author/list_number_of_readers_field.html.twig',
+            ])
+            ;
```
Comment (phansys) [inline tests/App/Admin/AuthorAdmin.php:42]: IIRC, this should be caught by PHP-CS-Fixer.
```suggestion
            ]);
```

File: tests/App/Entity/Author.php [hunk 2]
```diff
@@ -76,4 +92,19 @@ public function setAddress(?Address $address): void
...
+     * @return Collection<array-key, Book>
+     */
+    public function getBooks(): Collection
+    {
+        return $this->books;
+    }
+
+    public function getNumberOfReaders(): int
+    {
+        return array_sum($this->books->map(static function (Book $book) {
```
Comment (phansys) [inline tests/App/Entity/Author.php:106]:
```suggestion
        return array_sum($this->books->map(static function (Book $book): int {
```

File: tests/App/Admin/AuthorAdmin.php (hunk 1 above)
Comment (VincentLanglet) [reply tests/App/Admin/AuthorAdmin.php:42]: Do you want to make a PR on dev-kit for the `multiline_whitespace_before_semicolons` rule ?

Comment (phansys) [reply tests/App/Admin/AuthorAdmin.php:42]: I thought that it was already added. I'll provide a PR later.

File: src/Datagrid/Pager.php [hunk 3]
```diff
@@ -70,7 +70,26 @@ public function computeNbResult()
...
+        $identifierFieldNames = $query
+            ->getQueryBuilder()
+            ->getEntityManager()
+            ->getMetadataFactory()
+            ->getMetadataFor(current($query->getQueryBuilder()->getRootEntities()))
+            ->getIdentifierFieldNames();
+
+        // Paginator with fetchJoinCollection doesn't work with composite primary keys
+        // https://github.com/doctrine/orm/issues/2910
+        $paginator = new Paginator($query->getDoctrineQuery(), 1 === \count($identifierFieldNames));
```
Comment (franmomu) [inline src/Datagrid/Pager.php:90]: I guess it's unlikely, but `getDoctrineQuery` is not in the interface.

Comment (phansys) [reply src/Datagrid/Pager.php:90]: Indeed. IMO, we could add a call to `assert($query instanceof ...)` then.

Comment (VincentLanglet) [reply src/Datagrid/Pager.php]: Fixed

**Review by jordisala1991 — DISMISSED** [review]

File: tests/App/Entity/Author.php (hunk 2 above)
Comment (phansys) [reply tests/App/Entity/Author.php:106]: Could you please address this comment?

Comment (VincentLanglet) [reply tests/App/Entity/Author.php]: Done

**Review by phansys — APPROVED** [review]

**Pull request merged**